pkg/helper/helpers.go

```diff
@@ -11,34 +12,126 @@ import (
 	"k8s.io/client-go/util/retry"
 )
 
-// FindManifestConditionByIndex finds manifest condition by index in manifests
-func FindManifestConditionByIndex(index int32, conds []workapiv1.ManifestCondition) *workapiv1.ManifestCondition {
-	// Finds the cond conds that ordinal is the same as index
-	if conds == nil {
-		return nil
+// MergeManifestConditions return a new ManifestCondition array which merges the existing manifest
+// conditions and the new manifest conditions. Rules to match ManifestCondition between two arrays:
+// 1. match the manifest condition with the whole ManifestResourceMeta;
+// 2. if not matched, try to match with properties in ManifestResourceMeta other than ordinal
+// 3. otherwise, try to match with ordinal only in ManifestResourceMeta
+func MergeManifestConditions(conditions, newConditions []workapiv1.ManifestCondition) []workapiv1.ManifestCondition {
+	merged := []workapiv1.ManifestCondition{}
+
+	// build search indices
+	metaIndex := map[workapiv1.ManifestResourceMeta]workapiv1.ManifestCondition{}
+	metaWithoutOridinalIndex := map[string]workapiv1.ManifestCondition{}
+	ordinalIndex := map[int32]workapiv1.ManifestCondition{}
+
+	duplicateKeys := []string{}
+	for _, condition := range conditions {
+		metaIndex[condition.ResourceMeta] = condition
+		if key := manifestResourceMetaKey(condition.ResourceMeta); key != "" {
+			if _, exists := metaWithoutOridinalIndex[key]; exists {
+				duplicateKeys = append(duplicateKeys, key)
+			} else {
+				metaWithoutOridinalIndex[key] = condition
+			}
+		}
+		ordinalIndex[condition.ResourceMeta.Ordinal] = condition
 	}
-	for i, cond := range conds {
-		if index == cond.ResourceMeta.Ordinal {
-			return &conds[i]
+
+	// remove key from index if it is not unique
+	for _, key := range duplicateKeys {
+		delete(metaWithoutOridinalIndex, key)
+	}
+
+	// try to match and merge manifest conditions
+	for _, newCondition := range newConditions {
+		// match with ResourceMeta
+		condition, ok := metaIndex[newCondition.ResourceMeta]
+
+		// match with properties in ResourceMeta other than ordinal if not found yet
+		if !ok {
+			condition, ok = metaWithoutOridinalIndex[manifestResourceMetaKey(newCondition.ResourceMeta)]
 		}
+
+		// match with ordinal in ResourceMeta if not found yet
```

Comment:
if this happens, then I don't think you should consider them to be same.  Instead, just use the new condition.  I don't think ordinal is strong enough on its own to merit "sameness"